docs: add workaround to known issue loading missing policies from other namespaces (#23909)
hashicorp/vault: website/content/partials/known-issues/internal-error-namespace-missing-policy.mdx

- 1.14.4 and 1.14.5
- 1.15.0 and 1.15.1

A fix will be released in Vault 1.15.2, 1.14.6, and 1.13.10.

### Workaround

> During authentication, Vault derives inherited policies based on the groups an

Reviewer: ```suggestion

During authentication, Vault derives inherited policies based on the groups an
```

Add whitespace after headings for easier scanning in the markdown

As an example, consider the following userpass auth method failure. The error is
due to the fact that Vault expects a group policy under the namespace that does 
not exist.

<CodeBlockConfig hideClipboard>

```shell-session
# Failed login
$ vault login -method=userpass username=user1 password=123
Error authenticating: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/userpass/login/user1
Code: 500. Errors:

* internal error
```

</CodeBlockConfig>

To confirm the problem is a missing policy, start by identifying the relevant
entity and group IDs:

<CodeBlockConfig hideClipboard>

```shell-session
$ vault read -format=json identity/entity/name/user1 | \
  jq '{"entity_id": .data.id, "group_ids": .data.group_ids} '
{
  "entity_id": "420c82de-57c3-df2e-2ef6-0690073b1636",
  "group_ids": [
    "6cb152b7-955d-272b-4dcf-a2ed668ca1ea"
  ]
}
```

</CodeBlockConfig>

Use the group ID to fetch the relevant policies for the group under the `ns1`
namespace:

<CodeBlockConfig hideClipboard>

```shell-session
$ vault read -format=json -namespace=ns1 \
  identity/group/id/6cb152b7-955d-272b-4dcf-a2ed668ca1ea | \
  jq '.data.policies'
[
  "group_policy"
]
```

> </CodeBlockConfig>
> Now that we know Vault is looking for a policy called `group_policy`, we can

Reviewer: ```suggestion
</CodeBlockConfig>

Now that we know Vault is looking for a policy called `group_policy`, we can
```

The content parser sometimes gets tripped up when rendering the page if tag elements (like `<CodeBlockConfig>` ) are too close to other content. As a best practice, make sure to have a newline before and after any tag elements.

- 1.14.4 and 1.14.5
- 1.15.0 and 1.15.1

A fix will be released in Vault 1.15.2, 1.14.6, and 1.13.10.

### Workaround

> During authentication, Vault will derive any additional policies that should be inherited based on any groups an entity is a member of. If Vault attempts to attach a policy to a token from a different namespace than the one handling the authentication, _and_ that policy does not exist, Vault will return an internal error. The way to resolve this issue is to either add the policy, or to delete the group policy mapping referring to the non-existent policy.

Reviewer: ```suggestion
During authentication, Vault derives inherited policies based on the groups an
entity belongs to. Vault returns an internal error when attaching the derived
policy to a token when:

1. the token belongs to a different namespace than the one handling
   authentication, and
1. the derived policy does not exist under the namespace.

You can resolve the error by adding the policy to the relevant namespace or
deleting the group policy mapping that uses the derived policy.
```

Style correction: write in active voice

> Below is an example of a userpass auth method that is failing because of a
> missing group policy from another namespace.

Reviewer: ```suggestion
As an example, consider the following userpass auth method failure. The error is
likely due to the fact that Vault expects a group policy under the namespace for
`user1` that does not exist.
```
Style correction: write in active voice, make all participating elements explicit

> ```

Reviewer: ````suggestion
<CodeBlockConfig hideClipboard>

```shell-session
````

# Failed login
$ vault login -method=userpass username=user1 password=123
Error authenticating: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/userpass/login/user1
Code: 500. Errors:

> * internal error

Reviewer: ````suggestion
* internal error
```

</CodeBlockConfig>
````

> # Identify the entity and group IDs
> $ vault read -format=json identity/entity/name/user1 | jq '{"entity_id": .data.id, "group_ids": .data.group_ids} '
> {
>   "entity_id": "420c82de-57c3-df2e-2ef6-0690073b1636",
>   "group_ids": [
>     "6cb152b7-955d-272b-4dcf-a2ed668ca1ea"
>   ]
> }
> 

Reviewer: ````suggestion
To confirm the problem is a missing policy, start by identifying the relevant
entity and group IDs:

<CodeBlockConfig hideClipboard>

```shell-session
$ vault read -format=json identity/entity/name/user1 | \
  jq '{"entity_id": .data.id, "group_ids": .data.group_ids} '
{
  "entity_id": "420c82de-57c3-df2e-2ef6-0690073b1636",
  "group_ids": [
    "6cb152b7-955d-272b-4dcf-a2ed668ca1ea"
  ]
}
```

</CodeBlockConfig>

````

Style correction: avoid multiple steps per code block, introduce what the example code is doing

# Fetch the group from the other namespace

> $ vault read -format=json -namespace=ns1 identity/group/id/6cb152b7-955d-272b-4dcf-a2ed668ca1ea | jq '.data.policies'

Reviewer: Question: how would the reader know to specify `ns1` as the namespace? Are we assuming that's the namespace that `uesr1` belongs to? Or are we assuming it is the namespace under which the original call was made?

Author: The operator would need to know that the group (id: 6cb152b7-955d-272b-4dcf-a2ed668ca1ea ) was created in a certain namespace.

Reviewer: Is there a step we can add to the troubleshooting example to show folks how to find/confirm the relevant namespace. Either with the group ID or the username? Given that they need the namespace for the subsequent steps, it would be helpful to show folks how to identify the relevant namespace.

Author: There's not an easy way to find out what namespace the group or policy originated. The operator can do a check against different namespaces by looking for the group ID like this:
```
vault read -format=json -namespace=ns1 \
  identity/group/id/some_id
```
If group data is returned, they'll know it's the correct namespace.

> # Fetch the group from the other namespace
> $ vault read -format=json -namespace=ns1 identity/group/id/6cb152b7-955d-272b-4dcf-a2ed668ca1ea | jq '.data.policies'
> [
>   "group_policy"
> ]
> 

Reviewer: ````suggestion
Use the group ID to fetch the relevant policies for `user1` under the `ns1`
namespace:

<CodeBlockConfig hideClipboard>

```shell-session
$ vault read -format=json -namespace=ns1 \
  identity/group/id/6cb152b7-955d-272b-4dcf-a2ed668ca1ea | \
  jq '.data.policies'
[
  "group_policy"
]
```

</CodeBlockConfig>
````

> # The only policy in this namespace is default (group_policy is missing)
> $ vault policy list -namespace=ns1
> default

Reviewer: ````suggestion
Now that we know Vault is looking for a policy called `group_policy`, we can
check whether that policy exists under the `ns1` namespace:

<CodeBlockConfig hideClipboard>

```shell-session
$ vault policy list -namespace=ns1
default
```

</CodeBlockConfig>

The only policy in the `ns1` namespace is `default`, which confirms that the
missing policy (`group_policy`) is causing the error.
````

> # Option 1: Remove the missing policy from the group
> $ vault write -namespace=ns1 identity/group/id/6cb152b7-955d-272b-4dcf-a2ed668ca1ea name="test" policies="default"
> 
> # OR Option 2: Create the missing policy
> $ vault policy write -namespace=ns1 group_policy - << EOF
>     path "secret/data/*" {
> 	    capabilities = ["create", "update"]
>     }
> EOF
> 
> # Login should succeed now
> $ vault login -method=userpass username=user1 password=123`
> ```

Reviewer: ````suggestion

To fix the problem, we can either remove the missing policy from the
`6cb152b7-955d-272b-4dcf-a2ed668ca1ea` group or create the missing policy under
the `ns1` namespace.

<Tabs>

<Tab heading="Remove the group policy">

To remove `group_policy` from group ID `6cb152b7-955d-272b-4dcf-a2ed668ca1ea`,
use the `vault write` command to set the applicable policies to just include
`default`:

```shell-session
$ vault write                                             \
  -namespace=ns1                                          \
  identity/group/id/6cb152b7-955d-272b-4dcf-a2ed668ca1ea  \
  name="test"                                             \
  policies="default"
```

</Tab>

<Tab heading="Add the policy to ns1">

To create the missing policy, use `vault policy write` and define the
appropriate capabilities:

```shell-session
$ vault policy write -namespace=ns1 group_policy - << EOF
    path "secret/data/*" {
	    capabilities = ["create", "update"]
    }
EOF
```

</Tab>
</Tabs>

Verify the fix by re-running the login command:

<CodeBlockConfig hideClipboard>

```shell-session
$ vault login -method=userpass username=user1 password=123`
```

</CodeBlockConfig>
````

Style correction: make key elements easy to scan, avoid documenting in code comments, use one command per code block

- 1.14.4 and 1.14.5
- 1.15.0 and 1.15.1

A fix will be released in Vault 1.15.2, 1.14.6, and 1.13.10.

### Workaround

> During authentication, vault adds all policies that can be applied to an entity.
> This includes policies in separate namespaces. If vault attempts to add a policy
> from another namespace _and_ this policy does not exist, vault will return an
> internal error. The way to resolve this issue is to either add the policy, or to
> delete the policy.

Reviewer 2: ```suggestion
During authentication, Vault will derive any additional policies that should be inherited based on any groups an entity is a member of. If Vault attempts to attach a policy to a token from a different namespace than the one handling the authentication, _and_ that policy does not exist, Vault will return an internal error. The way to resolve this issue is to either add the policy, or to delete the group policy mapping referring to the non-existent policy.
```
